### ISSUE TITLE
passing the "c" flag to search() prevents it from finding a match

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6735]
**Describe the bug**

Passing the `c` flag to `search()` might prevent it from finding a match.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        let lines =<< trim END
            echo "foo"
            " bar
        END
        setline(1, lines)
        set ft=vim
        syn on
        search('"', 'cW', 0, 0, {->
            synstack('.', col('.'))
            ->map({_, v -> synIDattr(v, 'name')})->match('\ccomment') == -1})
    EOF
    )

The cursor didn't move; it's still on the first column of the first line.

**Expected behavior**

The cursor moved on the first uncommented quote which is on the first column of the second line.

**Environment**

 - Vim version: 8.2 Included patches: 1-1484
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: XTerm(358)

**Additional context**

The skip expression tells `search()` to skip any quote which is not highlighted as a comment:
```vim
synstack('.', col('.'))->map({_, v -> synIDattr(v, 'name')})->match('\ccomment') == -1
```
I would expect the cursor to jump on the last quote, because it's the only one for which the expression is false; i.e. the quote is highlighted as a comment.

I guess that the expression is evaluated on the first quote:

    echo "foo"
         ^

It evaluates to 0, because the quote is not highlighted as a comment.  I would then expect `search()` to skip the first quote, and evaluate the expression on the second one; then later on the third one.  But it doesn't do that.  I don't know whether it's working as intended, but even if it is, I find this unexpected.

---

The issue disappears if the `c` flag is removed:
```vim
vim9script
let lines =<< trim END
    echo "foo"
    " bar
END
setline(1, lines)
set ft=vim
syn on
search('"', 'W', 0, 0, {->
    synstack('.', col('.'))
    ->map({_, v -> synIDattr(v, 'name')})->match('\ccomment') == -1})
```
---

But for what I'm trying to do, I need the `c` flag.  I want to find any commented quote, including one which is already under the cursor.  Removing `c` would prevent `search()` to find the first quote in this text:

```vim
" comment
echo "foo"
" bar
```
Test:
```vim
vim9script
let lines =<< trim END
    " comment
    echo "foo"
    " bar
END
setline(1, lines)
set ft=vim
syn on
search('"', 'W', 0, 0, {->
    synstack('.', col('.'))
    ->map({_, v -> synIDattr(v, 'name')})->match('\ccomment') == -1})
```
Here, the cursor moved on the last commented quote, and ignored the first commented one because `c` was removed from the flags.

**Comment by lacygoill** [comment]
> But for what I'm trying to do, I need the c flag.

I found another way.  A regular `:s` command is much faster than what I was trying to do anyway.  Maybe it's working as intended, I'm not sure.

**Comment by lacygoill** [comment]
Fixed by the patch [8.2.5152](https://github.com/vim/vim/releases/tag/v8.2.5152).
```vim
vim9script
var lines =<< trim END
    echo "foo"
    " bar
END
lines->setline(1)
&filetype = 'vim'
syntax on
search('"', 'cW', 0, 0,
    () => synstack('.', col('.')) ->map((_, v) => synIDattr(v, 'name'))->match('\ccomment') == -1)
```
Thank you @zeertzjq
